plugins/bnr-dp/fu-bnr-dp-firmware.c

```diff
@@ -0,0 +1,611 @@
+	 * could technically be worked around by resetting the active boot counter */
+	if (active_boot_counter == G_MAXUINT32) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_WRITE,
+				    "update count exhausted");
+		return FALSE;
+	}
+
+	image = fu_firmware_get_bytes(firmware, error);
+	header = fu_struct_bnr_dp_payload_header_parse(g_bytes_get_data(image, NULL),
+						       g_bytes_get_size(image),
+						       FU_BNR_DP_FIRMWARE_HEADER_OFFSET,
+						       error);
+
+	/* check that the current CRC was correct */
+	crc = fu_crc16(FU_CRC_KIND_B16_BNR,
+		       header->data,
+		       FU_STRUCT_BNR_DP_PAYLOAD_HEADER_SIZE - sizeof(crc));
+	if (fu_struct_bnr_dp_payload_header_get_crc(header) != crc) {
+		g_set_error(
+		    error,
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_INVALID_FILE,
+		    "CRC mismatch in firmware binary header (header specified: 0x%X, actual: 0x%X)",
+		    fu_struct_bnr_dp_payload_header_get_crc(header),
+		    crc);
+		return FALSE;
+	}
+
+	/* set new counter */
+	g_info("incrementing boot counter: %u => %u", active_boot_counter, active_boot_counter + 1);
+	fu_struct_bnr_dp_payload_header_set_counter(header, active_boot_counter + 1);
+
+	/* clear CRC error flag if set for some reason */
+	fu_struct_bnr_dp_payload_header_set_flags(
+	    header,
+	    fu_struct_bnr_dp_payload_header_get_flags(header) & ~FU_BNR_DP_PAYLOAD_FLAGS_CRC_ERROR);
+
+	/* update checksum */
+	crc = fu_crc16(FU_CRC_KIND_B16_BNR,
+		       header->data,
+		       FU_STRUCT_BNR_DP_PAYLOAD_HEADER_SIZE - sizeof(crc));
+
+	fu_struct_bnr_dp_payload_header_set_crc(header, crc);
+
+	patch = g_bytes_new(header->data, header->len);
+	fu_firmware_add_patch(firmware, FU_BNR_DP_FIRMWARE_HEADER_OFFSET, patch);
+
+	return TRUE;
+}
+
+/* do checks that can only be done with data from an opened device */
+gboolean
+fu_bnr_dp_firmware_check(FuBnrDpFirmware *self,
+			 const FuStructBnrDpFactoryData *factory_data,
+			 const FuStructBnrDpPayloadHeader *active_header,
+			 const FuStructBnrDpPayloadHeader *fw_header,
+			 GError **error)
+{
+	FuFirmware *firmware = FU_FIRMWARE(self);
+	guint64 active_version = 0;
+	guint64 fw_version = 0;
+	g_autofree gchar *active_version_str = NULL;
+	g_autofree gchar *fw_version_str = NULL;
+	guint32 product_num;
+	guint16 compat_id;
+
+	/* compare versions */
+	if (!fu_bnr_dp_version_from_header(active_header, &active_version, error))
+		return FALSE;
+	if (!fu_bnr_dp_version_from_header(fw_header, &fw_version, error))
+		return FALSE;
+	active_version_str = fu_bnr_dp_version_to_string(active_version);
+	fw_version_str = fu_bnr_dp_version_to_string(fw_version);
+	if (fu_firmware_get_version_raw(firmware) != fw_version) {
+		g_set_error(
+		    error,
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_INVALID_DATA,
+		    "versions in firmware XML header (%s) and binary payload (%s) are inconsistent",
+		    fu_firmware_get_version(firmware),
+		    fw_version_str);
+		return FALSE;
```

Comment:
from experience in other plugins; I'd ask when this happens do you want an escape hatch without recompiling?  If so look at how some other plugins how they can use the `--force` argument to override this.